cocos/physics/bullet/character-controllers/bullet-character-controller.ts

```diff
@@ -0,0 +1,249 @@
+        if (!this._impl) return;
+        bt.CharacterController_setStepOffset(this._impl, value);
+    }
+
+    setSlopeLimit (value: number): void {
+        if (!this._impl) return;
+        bt.CharacterController_setSlopeLimit(this._impl, degreesToRadians(value));
+    }
+
+    setDetectCollisions (value: boolean): void {
+        if (!this._impl) return;
+        bt.CharacterController_setCollision(this.impl, value);
+    }
+
+    setOverlapRecovery (value: boolean): void {
+        if (!this._impl) return;
+        bt.CharacterController_setOverlapRecovery(this.impl, value);
+    }
+
+    onGround (): boolean {
+        return (this._btCollisionFlags & (1 << 2)) > 0;//btControllerCollisionFlag::Enum::BULLET_CONTROLLER_COLLISION_DOWN
+    }
+
+    syncSceneToPhysics (): void {
+        const node = this.characterController.node;
+        if (node.hasChangedFlags) {
+            //only sync scale from scene node to physics
+            if (node.hasChangedFlags & TransformBit.SCALE) this.syncScale();
+        }
+    }
+
+    syncPhysicsToScene (): void {
+        this.getPosition(v3_0);
+        v3_0.subtract(this._comp.scaledCenter);
+        this._comp.node.setWorldPosition(v3_0);
+    }
+
+    syncScale () {
+        this.updateScale();
+    }
+
+    move (movement: IVec3Like, minDist: number, elapsedTime: number) {
+        if (!this._isEnabled) { return; }
+        const movementBT = BulletCache.instance.BT_V3_0;
+        bt.Vec3_set(movementBT, movement.x, movement.y, movement.z);
+        this._btCollisionFlags = bt.CharacterController_move(this.impl, movementBT, minDist, elapsedTime);
+    }
+
+    setGroup (v: number): void {
+        if (v !== this._collisionFilterGroup) {
+            this._collisionFilterGroup = v;
+            this._dirty = true;
+        }
+    }
+
+    getGroup (): number {
+        return this._collisionFilterGroup;
+    }
+
+    addGroup (v: number): void {
+        this._collisionFilterGroup |= v;
+        this._dirty = true;
+    }
+
+    removeGroup (v: number): void {
+        this._collisionFilterGroup &= ~v;
+        this._dirty = true;
+    }
+
+    setMask (v: number): void {
+        if (v !== this._collisionFilterMask) {
+            this._collisionFilterMask = v;
+            this._dirty = true;
+        }
+    }
+
+    getMask (): number {
+        return this._collisionFilterMask;
+    }
+
+    addMask (v: number): void {
+        this._collisionFilterMask |= v;
+        this._dirty = true;
+    }
+
+    removeMask (v: number): void {
+        this._collisionFilterMask &= ~v;
+        this._dirty = true;
+    }
+
+    updateEventListener () {
+        (PhysicsSystem.instance.physicsWorld as BulletWorld).updateNeedEmitCCTEvents(this.characterController.needCollisionEvent);
+    }
+
+    // update group and mask by re-adding cct to physics world
+    updateDirty () {
+        if (this._dirty) {
+            (PhysicsSystem.instance.physicsWorld as BulletWorld).removeCCT(this);
```

Comment:
The cost seems a bit heavy. Just update group and mask settings?

Reply:
bullet needs to readd the cct to change its mask. https://pybullet.org/Bullet/phpBB3/viewtopic.php?t=8464